src/main/scala/akka/contrib/process/BlockingProcess.scala

```diff
@@ -102,17 +99,18 @@ class BlockingProcess(
 
   private val process = startProcess()
 
+  // This shutdown hook is required because the stdio actors will be blocked on a read. The only
+  // reliable means of unblocking them is to destroy the thing they're listening to.
+  context.actorOf(ShutdownHook.props(() => process.destroy()), "process-destroyer")
```

Comment:
>No, it is not guaranteed - just highly unlikely.

I propose to make it deterministic, seems like a bug waiting to happen otherwise.

>One is for the regular use-case of terminating a process - this will be done in a known thread-safe manner. 

But it will mean that the shutdownhook will also call `destroy()` on exit.

>The other use-case is to catch actor system shutdown and tear everything down no matter what - it actually doesn't matter whether other threads for the parent actor hierarchy barf or not - the actor system is on its way out.

That doesn't solve the case of JVM crash.

>We'd have to resort to native code.

Sounds like a nightmare.

>Anyhow, all that I can say is that, at least for now, this works. It is the only thing I've found to >work, and I've been playing with this for a while. Please recommend a solution though. I'm all >ears. The JDK is screwed when it comes to unblocking streams.

I'd suggest to use the same code path to destroy the process no matter if it is on request (Destroy) or as actors are stopped (shutdownhook actor) and to log if Process.destroy() fails, since it could indicate that the process is still running, or other issues, like the aforementioned thread safety issue.

Does that make sense?